src/tech/v3/dataset/impl/column_index_structure.clj

```diff
@@ -0,0 +1,85 @@
+(ns tech.v3.dataset.impl.column-index-structure
+  (:import [java.util TreeMap LinkedHashMap]
+           [tech.v3.datatype ListPersistentVector])
+  (:require [tech.v3.protocols.column :as col-proto]
+            [tech.v3.dataset.impl.column-base :refer [column-datatype-categorical?]]
+
+            [tech.v3.datatype :refer [elemwise-datatype clone ->buffer]]
+            [tech.v3.datatype.argops :refer [arggroup]]
+            [tech.v3.datatype.casting :refer [datatype->object-class]]
+            [clojure.set :refer [difference]]))
+
+
+(extend-type TreeMap
+  col-proto/PIndexStructure
+  (select-from-index [index-structure mode selection-spec]
+    (case mode
+      :pick
+      (let [^TreeMap new-index-structure (.clone ^TreeMap index-structure)
+            s (difference (set (.keySet new-index-structure)) (set selection-spec))]
+        (doseq [k s]
+          (.remove new-index-structure k))
+        new-index-structure)
+      :slice
+      (let [{from            :from
+             from-inclusive? :from-inclusive?
+             to              :to
+             to-inclusive?   :to-inclusive?} selection-spec]
+        (.subMap ^TreeMap index-structure
+                 from
+                 (if (nil? from-inclusive?) true from-inclusive?)
+                 to
+                 (if (nil? to-inclusive?) true to-inclusive?))))))
+
+
+(extend-type LinkedHashMap
+  col-proto/PIndexStructure
+  (select-from-index [index-structure mode selection-spec]
+    (case mode
+      :pick
+      (let [^LinkedHashMap new-index-structure (.clone ^LinkedHashMap index-structure)
+            s (difference (set (.keySet new-index-structure)) (set selection-spec))]
+        (doseq [k s]
+          (.remove new-index-structure k))
+        new-index-structure))))
+
+
+;; https://github.com/techascent/tech.ml.dataset/blob/master/src/tech/v3/dataset/impl/column.clj#L41
+(defn categorical? [dtype]
+  (boolean (#{:string :keyword :symbol} dtype)))
+
+
+(defn build-value-to-index-position-map [column-data]
+  (let [idx-map (arggroup column-data)
+        vals->list-persistent-vector (fn [_ list-data]
+                                       (-> (clone list-data) ->buffer ListPersistentVector.))]
+    (.replaceAll idx-map (reify java.util.function.BiFunction
+                           (apply [this k v]
+                             (vals->list-persistent-vector k v))))
+    idx-map))
+
+
+(defmulti make-index-structure
+  "Returns an index structure based on the type of data in the column."
+  (fn [data metadata]
+    (let [data-dtype (-> data elemwise-datatype)
+          data-klass (-> data-dtype datatype->object-class)]
+      (if (contains? metadata :categorical?)
+        (if (:categorical? metadata)
+          ::categorical
+          data-klass)
+        (if (column-datatype-categorical? data-dtype)
+          ::categorical
+          data-klass)))))
+
```

Comment:
What is the intention of using data-klass here as opposed to the column's elemwise datatype?

Reply:
Ahhh good question. The thinking here was that this might make the extensibility a bit easier. The multimethod matching is sensitive to java class hierarchies, so dispatching the class can make it simpler to match. Here's an example from the [GIS example](https://github.com/scicloj/sci-fu/blob/ethan/tmd-column-extensibility/projects/index-experiments/src/index_experiments/tmd_column_extension_demo.clj#L90-L98):
```clj
(defmethod tech.v3.dataset.impl.column-index-structure/make-index-structure Geometry
  [data _]
  (let [tree ^STRtree (STRtree.)]
    (doseq [[index-position geometry] (map-indexed vector data)]
      (.insert tree
               (.getEnvelopeInternal geometry)
               {:index-position index-position
                :prepared-geometry (PreparedGeometryFactory/prepare geometry)}))
    tree))
```
Whatever kind of geometry is in the column, the method will match b/c those geometries all extend `Geometry`. I think since I wrote this, I realized there may be some other approaches. For example, we could dispatch the datatype and expect the end user extending tmd to use Clojure's custom hierarchies, or maybe they could also use tech.datatype's aliasing function?

What do you think?

Reply:
Well, if you want to use an actual class like that I think it is reasonable but the check is more specific.  Something like `if there isn't a datatype registered for this class, use the class else use the registered datatype`.  You know things aren't registered if they return ':object' and there are more specific checks also in tech.v3.datatype.casting.  You could register a namespaced datatype for geometry as another option. 

Also, the dispatch method should *always* return the datatype if registered or the class itself.  The default implementation should be checking for :categorical?, not the multimethod dispatch function.